tests/MaxMind/Test/ReportTransactionTest.php

```diff
@@ -0,0 +1,320 @@
+     *
+     * @param array $req
+     */
+    public function testMissingRequiredFieldsWithoutValidation($req)
+    {
+        $this->createReportTransactionRequest(
+            $req,
+            0,
+            ['validateInput' => false]
+        )->reportTransaction($req);
+    }
+
+    public function requestsMissingRequiredFields()
+    {
+        return [
+            'Missing ip_address' => [
+                ['tag' => 'not_fraud'],
+            ],
+            'Missing tag' => [
+                ['ip_address' => '1.2.3.4'],
+            ],
+            'Null ip_address' => [
+                array_merge(Data::minimalRequest(), ['ip_address' => null]),
+            ],
+            'Null tag' => [
+                array_merge(Data::minimalRequest(), ['tag' => null]),
+            ],
+        ];
+    }
+
+    /**
+     * @expectedException \MaxMind\Exception\InvalidInputException
+     * @expectedExceptionMessage Must have keys
+     */
+    public function testUnknownKey()
+    {
+        $req = array_merge(
+            Data::minimalRequest(),
+            ['unknown' => 'some_value']
+        );
+        $this->createReportTransactionRequest(
+            $req,
+            0
+        )->reportTransaction($req);
+    }
+
+    /**
+     * @expectedException \MaxMind\Exception\InvalidInputException
+     * @expectedExceptionMessage chargeback_code must be a string
+     *
+     * @dataProvider notStringTypes
+     *
+     * @param mixed $chargebackCode
+     */
+    public function testInvalidChargebackCodes($chargebackCode)
+    {
+        $req = array_merge(
+            Data::minimalRequest(),
+            ['chargeback_code' => $chargebackCode]
+        );
+        $this->createReportTransactionRequest(
+            $req,
+            0
+        )->reportTransaction($req);
+    }
+
+    /**
+     * @expectedException \MaxMind\Exception\InvalidInputException
+     * @expectedExceptionMessage notes must be a string
+     *
+     * @dataProvider notStringTypes
+     *
+     * @param mixed $notes
+     */
+    public function testInvalidNotes($notes)
+    {
+        $req = array_merge(Data::minimalRequest(), ['notes' => $notes]);
+        $this->createReportTransactionRequest(
+            $req,
+            0
+        )->reportTransaction($req);
+    }
+
+    /**
+     * @expectedException \MaxMind\Exception\InvalidInputException
+     * @expectedExceptionMessage transaction_id must be a string
+     *
+     * @dataProvider notStringTypes
+     *
+     * @param mixed $transactionId
+     */
+    public function testInvalidTransactionIds($transactionId)
+    {
+        $req = array_merge(
+            Data::minimalRequest(),
+            ['transaction_id' => $transactionId]
+        );
+        $this->createReportTransactionRequest(
+            $req,
+            0
+        )->reportTransaction($req);
+    }
+
+    public function notStringTypes()
+    {
+        return [
+            [1],
```

Comment:
If we've bothered to create a data provider, we might as well add some other data types. Maybe an array or an object?